apikey.go

```diff
@@ -0,0 +1,188 @@
+// license that can be found in the LICENSE file.
+
+package kes
+
+import (
+	"crypto"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/pem"
+	"errors"
+	"io"
+	"math/big"
+	"strings"
+	"time"
+)
+
+// APIKey is an object that can generate a private/public key pair.
+type APIKey interface {
+	// Public returns the public key that coresponds to
+	// the public key.
+	Public() crypto.PublicKey
+
+	// Private returns the private key that correspons
+	// to the public key.
+	Private() crypto.PrivateKey
+
+	// Identity returns the Identity of the public key.
+	//
+	// The identity is the cryptographic fingerprint of
+	// the raw DER-encoded public key as present in a
+	// corresponding X509 cerificate.
+	Identity() Identity
+
+	// String returns the APIKey's textual representation.
+	String() string
+}
+
+// GenerateAPIKey generates a new random API key. If rand
+// is nil, crypto/rand.Reader is used.
+func GenerateAPIKey(rand io.Reader) (APIKey, error) {
+	pub, priv, err := ed25519.GenerateKey(rand)
+	if err != nil {
+		return nil, err
+	}
+	id, err := ed25519Identity(pub)
+	if err != nil {
+		return nil, err
+	}
+	return &apiKey{
+		key:      priv,
+		identity: id,
+	}, nil
+}
+
+// ParseAPIKey parses a formatted APIKey and returns the
+// value it represents.
+func ParseAPIKey(s string) (APIKey, error) {
+	const (
+		Header      = "kes:v1:"
+		Ed25519Type = 0
+	)
+	if !strings.HasPrefix(s, Header) {
+		return nil, errors.New("kes: invalid API key: missing 'kes:v1:' prefix")
+	}
+	b, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(s, Header))
+	if err != nil {
+		return nil, err
+	}
+	if len(b) != 1+ed25519.SeedSize {
+		return nil, errors.New("kes: invalid API key: invalid length")
+	}
+	if b[0] != Ed25519Type {
+		return nil, errors.New("kes: invalid API key: unsupported type")
+	}
+	key := ed25519.NewKeyFromSeed(b[1:])
+	id, err := ed25519Identity(key[ed25519.SeedSize:])
+	if err != nil {
+		return nil, err
+	}
+	return &apiKey{
+		key:      key,
+		identity: id,
+	}, nil
+}
+
+// CertificateOption is a function modifying the passed *x509.Certificate.
+type CertificateOption func(*x509.Certificate)
+
+// GenerateCertificate generates a new tls.Certificate from the APIKey's
+// private and public key. The certificate can be customized by specifying
+// one or multple CertificateOptions.
+func GenerateCertificate(key APIKey, options ...CertificateOption) (tls.Certificate, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			CommonName: key.Identity().String(),
+		},
+		NotBefore: time.Now().UTC(),
+		NotAfter:  time.Now().UTC().Add(90 * 24 * time.Hour),
```

Comment:
Maybe note in the function that default is 90 days. I suppose it can be altered by an option.